torch/_vmap_internals.py

```diff
@@ -34,19 +34,65 @@
     'output (got {num_outputs} outputs) of {fn}.'
 )
 
+EXPECTED_IN_DIMS_TO_BE_INT_OR_TUPLE = (
+    'vmap({fn}, in_dims={in_dims}, ...): expected `in_dims` to be int or tuple, '
+    'got: {actual_type}.'
+)
+
+IN_DIMS_AND_NUM_INPUTS_MISMATCH = (
+    'vmap({fn}, in_dims={in_dims}, ...)(<inputs>): expected one `in_dim` per '
+    'input (got {num_inputs} inputs) of {fn}'
+)
 
-# Checks that all args have the same batch dim size.
-def _validate_and_get_batch_size(args):
-    batch_sizes = [arg.size(0) for arg in args]
+IN_DIMS_MUST_BE_FLAT_TUPLE = (
+    'vmap({fn}, in_dims={in_dims}, ...)(<inputs>): in_dims must be a flat '
+    'tuple containing ints and/or Nones. If you were trying to vmap over a '
+    'Tensor inside a Python collection in `inputs`, we do not yet support that.'
+)
+
+CANT_VMAP_A_NONTENSOR = (
+    'vmap({fn}, in_dims={in_dims}, ...)(<inputs>): Got in_dim={in_dim} for '
+    'input {idx}, but input {idx} is not a Tensor (got {arg_type}) so it '
+    'cannot be vmap\'ed over. If you were trying to vmap over a Tensor inside '
+    'a Python collection in `inputs`, we do not yet support that; otherwise, '
+    'use None as the respective in_dim for input {idx}.'
+)
+
+IN_DIM_NOT_IN_TENSOR = (
+    'vmap({fn}, in_dims={in_dims}, ...)(<inputs>): Got in_dim={in_dim} for '
+    'input {idx}, but input {idx} is a Tensor of dimensionality {tensor_dim} '
+    'so expected in_dim to satisfy 0 <= in_dim < {tensor_dim}.'
+)
+
+# Checks that all args-to-be-batched have the same batch dim size
+def _validate_and_get_batch_size(in_dims_as_tuple, args):
+    batch_sizes = [arg.size(in_dim) for in_dim, arg in zip(in_dims_as_tuple, args)
+                   if in_dim is not None]
     if batch_sizes and any([size != batch_sizes[0] for size in batch_sizes]):
         raise ValueError(REQUIRE_SAME_MAP_SIZE.format(sizes=batch_sizes))
     return batch_sizes[0]
 
+# Check compatibility of `in_dims` and `args`. More specifically, checks the following:
+# Wherever an in_dim is not None, then the corresponding index in args must be
+# a Tensor. Furthermore, tensor must have the `in_dim` (0 <= in_dim < tensor.dim())
```

Comment:
Type signature on this function (and the others) would be very helpful!

Reply:
RIP Python 2!!!

Reply:
I added type hints to all of the functions :D. We'll have to relax some of these in the future when we support accepting arbitrary nested python data structures, but these do make the code easier to read now.